### PR TITLE
laplace method

### LINKED ISSUE #760: Add method laplace_sample
Introduced in CmdStan 2.32. See CmdStanPy discussion: https://github.com/stan-dev/cmdstanpy/issues/649#issue-1549250952

### PR DESCRIPTION
**Pull request by jgabry**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Closes #760 
Builds off of PR #799 

Implements new method `CmdStanModel$laplace()` with new fitted model class `CmdStanLaplace`. 
Follows the design from @WardBrian in https://github.com/stan-dev/cmdstanr/issues/760#issuecomment-1516609693. 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting
(this will be you or your assignee, such as a university or company):
**Columbia University**

By submitting this pull request, the copyright holder is agreeing to
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by jgabry** [comment]
### Example usage

```r
file <- file.path(cmdstan_path(), "examples/bernoulli/bernoulli.stan")
mod <- cmdstan_model(file)
mod$print()

stan_data <- list(N = 10, y = c(0,1,0,0,0,0,0,0,0,1))
fit_mode <- mod$optimize(data = stan_data, jacobian = TRUE)
fit_laplace <- mod$laplace(data = stan_data, mode = fit_mode)
fit_laplace$summary()
fit_laplace$draws("theta")

# if mode isn't specified optimize is run internally first
fit_laplace <- mod$laplace(data = stan_data)
fit_laplace$summary()
```

**Comment by jgabry** [comment]
To-do note to self: avoid running Laplace if optimization fails, with informative error message

**Comment by jgabry** [comment]
I think this is ready for review. I've implemented everything that I can think of. Anything I need to do when adding a new method that I'm forgetting?

**Comment by jgabry** [comment]
Any ideas why everything is passing except with the WSL backend?

**Comment by jgabry** [comment]
I agree. Those are the names CmdStan uses, but I actually added the `draws` argument to `variational` in a recent commit to match Laplace. `output_samples` will still work but we can deprecate it eventually.

**Comment by avehtari** [comment]
I guess this should be done in CmdStan, but mentioning now here. It would be nice to be able to be able to parallelize the computation for the draws from the approximation similarly way as `$sampling()` can do. Often the optimization is fast, but also often obtaining 4000 draws and computing `lp__` and `lp_approx__` takes time that is similar to getting 1000 MCMC draws.

**Comment by avehtari** [comment]
Here's an example to do resample importance sampling
```
file <- file.path(cmdstan_path(), "examples/bernoulli/bernoulli.stan")
mod <- cmdstan_model(file)
mod$print()

stan_data <- list(N = 10, y = c(0,1,0,0,0,0,0,0,0,0))
fit_mode <- mod$optimize(data = stan_data, jacobian = TRUE)
# 4000 draws from the approximation
fit_laplace <- mod$laplace(data = stan_data, mode = fit_mode,
                           draws = 4000, refresh=0)
draws_laplace <- fit_laplace$draws(variables="theta")
# Compute Pareto smoothed importance sampling weights
(psis_laplace <- psis(log_ratios=fit_laplace$draws("lp__")-fit_laplace$draws("lp_approx__"), r_eff=1))
# Resample 4000 draws using simple resampling without resampling
draws_psis <- resample_draws(draws_laplace,
                             weights = exp(psis_laplace$log_weights),
                             ndraws = 3000,
                             method = 'simple')

summarize_draws(draws_laplace, default_summary_measures())
summarize_draws(draws_psis, default_summary_measures())

library(patchwork)
p1 <- mcmc_areas(draws_laplace) + xlim(c(0,0.7))
p2 <- mcmc_areas(draws_psis) + xlim(c(0,0.7))
p1 / p2
```

**Comment by avehtari** [comment]
Is it possible to get draws without evaluating `lp__` and `lp_approx__`? In the cases when I would trust the normal approximation, there is unnecessary but computationally potentially very costly evaluation of `lp__`.

**Comment by jgabry** [comment]
> I guess this should be done in CmdStan, but mentioning now here. It would be nice to be able to be able to parallelize the computation for the draws from the approximation similarly way as `$sampling()` can do. Often the optimization is fast, but also often obtaining 4000 draws and computing `lp__` and `lp_approx__` takes time that is similar to getting 1000 MCMC draws.

I agree. But yeah I think this is more of a CmdStan issue than a CmdStanR issue.

**Comment by jgabry** [comment]
> Is it possible to get draws without evaluating `lp__` and `lp_approx__`? In the cases when I would trust the normal approximation, there is unnecessary but computationally potentially very costly evaluation of `lp__`.

Not by changing anything in CmdStanR unfortunately. All we're doing is providing access to what CmdStan has already computed and written to CSV, so this would need to happen in CmdStan.

**Comment by jgabry** [comment]
> Here's an example to do resample importance sampling
> 
> ```
> file <- file.path(cmdstan_path(), "examples/bernoulli/bernoulli.stan")
> mod <- cmdstan_model(file)
> mod$print()
> 
> stan_data <- list(N = 10, y = c(0,1,0,0,0,0,0,0,0,0))
> fit_mode <- mod$optimize(data = stan_data, jacobian = TRUE)
> # 1000 draws from the approximation
> fit_laplace <- mod$laplace(data = stan_data, mode = fit_mode, refresh=0)
> draws_laplace <- fit_laplace$draws(variables="theta")
> # Compute Pareto smoothed importance sampling weights
> (psis_laplace <- psis(log_ratios=fit_laplace$draws("lp__")-fit_laplace$draws("lp_approx__"), r_eff=1))
> # Resample 1000 draws using simple resampling without resampling
> draws_psis <- resample_draws(draws_laplace,
>                              weights = exp(psis_laplace$log_weights),
>                              ndraws = 1000,
>                              method = 'simple')
> 
> summarize_draws(draws_laplace, default_summary_measures())
> summarize_draws(draws_psis, default_summary_measures())
> 
> library(patchwork)
> p1 <- mcmc_areas(draws_laplace) + xlim(c(0,0.7))
> p2 <- mcmc_areas(draws_psis) + xlim(c(0,0.7))
> p1 / p2
> ```

Cool, thanks! Maybe we should put this in the documentation somewhere.

File: R/fit.R [hunk 1]
```diff
@@ -707,21 +710,23 @@ CmdStanFit$set("public", name = "constrain_variables", value = constrain_variabl
...
-#' (also on the unconstrained space). It is exposed in the variational method
-#' for performing the checks described in Yao et al. (2018) and implemented in
-#' the \pkg{loo} package.
+#' For variational inference `lp_approx__` is the log density of the variational
+#' approximation to `lp__` (also on the unconstrained space). It is exposed in
+#' the variational method for performing the checks described in Yao et al.
+#' (2018) and implemented in the \pkg{loo} package.
+#'
+#' For Laplace approximation `lp_approx__` is the the unnormalized density of
+#' the Laplace approximation.
```
Comment (avehtari) [inline R/fit.R:719]: Add: It can be used to perform the same checks as in the case of the variational method described in Yao et al. (2018).

**Review by avehtari — COMMENTED** [review]
Looks good. Added one suggestion

File: R/fit.R (hunk 1 above)
Comment (jgabry) [reply R/fit.R]: Thanks, just pushed a commit adding this

**Comment by jgabry** [comment]
@andrjohns When you have time would you be able to check if you can figure out why this is failing on WSL only? Thank you!

**Comment by andrjohns** [comment]
Ack sorry for forgetting about this! I'll have time to dig in on Friday

**Comment by jgabry** [comment]
No worries, and thanks!

**Comment by andrjohns** [comment]
Just a heads up that I'm still working on this, I'm on an M1 Mac as my daily driver now so I'm just ironing out some kinks on getting WSL running in my windows VM

**Comment by jgabry** [comment]
Ok thanks for working on this @andrjohns

**Comment by andrjohns** [comment]
@jgabry ended up being a super minor issue of path handling, all good to go now!

**Comment by jgabry** [comment]
Awesome, thanks @andrjohns! Glad it turned out to be a simple fix. Everything is passing so I'll go ahead and merge.

**Pull request merged**